File: FuncComp/Parsing/Lexer.cs
using System;
using System.Collections.Generic;
using System.Net.NetworkInformation;

namespace FuncComp.Parsing
{
    public class Lexer
    {
        public IEnumerable<Token> Lex(string input)
        {
            var state = new State(input);

            while (!state.IsEof)
            {
                if (char.IsWhiteSpace(state.Current))
                {
                    state.Offset++;
                }
                else if (char.IsDigit(state.Current))
                {
                    yield return LexNumber(ref state);
                }
                else if (char.IsLetter(state.Current))
                {
                    yield return LexIdentifier(ref state);
                }
                else if (state.TryGetN(2, out var value) && value == "||")
                {
                    while (!state.IsEof && state.Current != '\n') state.Offset++;
                }
                else
                {
                    yield return LexToken(ref state);
                }
            }
        }

        private Token LexNumber(ref State state)
        {
            var start = state.Offset;

            while (!state.IsEof && char.IsDigit(state.Current)) state.Offset++;

            var value = state.Input.Substring(start, state.Offset - start);

            return new Token(TokenType.Number, value);
        }

        private Token LexIdentifier(ref State state)
        {
            var start = state.Offset;

            while (!state.IsEof && IsIdentifierChar(state.Current)) state.Offset++;

            var value = state.Input.Substring(start, state.Offset - start);

            return new Token(TokenType.Identifier, value);
        }

        private Token LexToken(ref State state)
        {
            var twoCharMatch = state.TryGetN(2, out var twoCharValue);

            if (twoCharMatch && (twoCharValue == "==" || twoCharValue == "~=" || twoCharValue == ">=" || twoCharValue == "<=" || twoCharValue == "->"))
            {
                state.Offset += 2;

                return new Token(TokenType.Other, twoCharValue);
            }

            var value = state.Current.ToString();
            state.Offset++;
            return new Token(TokenType.Other, value);
        }

        private static bool IsIdentifierChar(in char c) => char.IsLetter(c) || char.IsDigit(c) || c == '_';

        private struct State
        {
            public string Input { get; }

            public char Current => Get(Offset);
            public char Get(int offset) => Input[offset];
            public bool TryGetN(int length, out string value) => TryGetN(Offset, length, out value);
            public bool IsEof => IsOffsetEof(Offset);
            public bool IsOffsetEof(int offset) => offset >= Input.Length;

            public int Offset;

            public State(string input)
            {
                Input = input;
                Offset = 0;
            }

            public bool TryGetN(int offset, int length, out string value)
            {
                if (IsOffsetEof(offset + length - 1))
                {
                    value = "";
                    return false;
                }

                value = Input.Substring(offset, length);
                return true;
            }
        }
    }
}

File: FuncComp.Parsing/Lexer.fs
module FuncComp.Parsing.Lexer

open System;

type State = string * int

let private current ((input, offset) : State) = input.[offset]
let private next ((input, offset) : State) = (input, offset + 1)
let private isEof ((input, offset) : State) = offset >= input.Length

let private takeWhile (func : char -> bool) (state : State) =
    let rec nextUntil (state : State) =
        if isEof state then state
        else if current state |> func then nextUntil <| next state
        else state

    let (input, startOffset) = state
    let (_, endOffset) = nextUntil state

    (input.Substring(startOffset, endOffset - startOffset), (input, endOffset))

let private (|WhiteSpace|_|) (state : State) = if current state |> Char.IsWhiteSpace then next state |> Some else None
let private (|NumberStart|_|) (state : State) = if current state |> Char.IsDigit then state |> Some else None
let private (|IdentifierStart|_|) (state : State) = if current state |> Char.IsLetter then state |> Some else None
let private (|Literal|_|) (lit : string) ((input, offset) : State) =
    if offset + lit.Length >= input.Length then None
    else if lit = input.Substring(offset, lit.Length) then Some (lit, (input, offset + lit.Length))
    else None

let private isIdent c = Char.IsLetter(c) || Char.IsDigit(c) || c = '_'

let lex (input : string) : Token seq =
    let lexNumber (state : State) =
        let (str, state) = takeWhile Char.IsDigit state
        ([str |> int |> Token.Number], state)
    let lexIdent (state : State) =
        let (str, state) = takeWhile isIdent state
        ([str |> Token.Identifier], state)

    let rec lex (state : State) : Token seq =
        if isEof state then Seq.empty
        else match state with
             | WhiteSpace s -> lex s
             | NumberStart s ->
                    let (tokens, s) = lexNumber s
                    Seq.append tokens <| lex s
             | IdentifierStart s ->
                    let (tokens, state) = lexIdent state
                    Seq.append tokens <| lex state
             | Literal "||" (_, s) -> takeWhile (fun c -> c <> '\n') s |> snd |> lex
             | Literal "==" (v, s) -> Seq.append [Token.Other v] (lex s)
             | Literal "~=" (v, s) -> Seq.append [Token.Other v] (lex s)
             | Literal ">=" (v, s) -> Seq.append [Token.Other v] (lex s)
             | Literal "<=" (v, s) -> Seq.append [Token.Other v] (lex s)
             | Literal "->" (v, s) -> Seq.append [Token.Other v] (lex s)
             | s -> Seq.append [current s |> string |> Token.Other] (lex <| next state)

    lex (input, 0)
